### PR TITLE
[luci] Convert NCHW to NHWC (CIRCLEINPUT, CIRCLEOUTPUT)

### PR DESCRIPTION
**Pull request by jinevening**
This converts NCHW to NHWC for CIRCLEINPUT and CIRCLEOUTPUT

ONE-DCO-1.0-Signed-off-by: Hyukjin Jeong <hj1.jeong@samsung.com>

---
Related to: #5401
Draft PR: #5438

### CODE REVIEW

File: compiler/luci/pass/src/ConvertNCHWToNHWCPass.cpp [hunk 1]
```diff
@@ -55,7 +55,6 @@ class DataFormatAnnotation final : public loco::NodeAnnotation
 
 void set_data_format(loco::Node *node, const DataFormat &format)
 {
-  assert(node->annot<DataFormatAnnotation>() == nullptr);
```
Comment (seanshpark) [inline compiler/luci/pass/src/ConvertNCHWToNHWCPass.cpp:58]: This change doesn't seem to be related with the description.

Comment (jinevening) [reply compiler/luci/pass/src/ConvertNCHWToNHWCPass.cpp]: I've made a separate PR for this. Thanks.

File: compiler/luci/pass/src/ConvertNCHWToNHWCPass.cpp [hunk 2]
```diff
@@ -212,6 +212,58 @@ class ConvertNCHWToNHWC final : public luci::CircleNodeMutableVisitor<bool>
...
+  bool visit(luci::CircleInput *node)
+  {
+    const auto n = node->dim(0).value();
+    const auto c = node->dim(1).value();
+    const auto h = node->dim(2).value();
+    const auto w = node->dim(3).value();
+
+    node->dim(1).set(h);
+    node->dim(2).set(w);
+    node->dim(3).set(c);
```
Comment (llFreetimell) [inline compiler/luci/pass/src/ConvertNCHWToNHWCPass.cpp:224]: When considering unknown dimensions, following codes

```suggestion
    const auto n = node->dim(0);
    const auto c = node->dim(1);
    const auto h = node->dim(2);
    const auto w = node->dim(3);

    node->dim(1) = h;
    node->dim(2) = w;
    node->dim(3) = c;
```

 or adding assert would be better.

**Comment by seanshpark** [comment]
Input shape something like (-1, 10), (10, -1) also exists. @jinevening can you check with these?

**Comment by jinevening** [comment]
> @jinevening can you check with these?

Sure, I'll add a check code that prevents converting dynamic-shaped models.

File: compiler/luci/pass/src/ConvertNCHWToNHWCPass.cpp (hunk 2 above)
Comment (jinevening) [reply compiler/luci/pass/src/ConvertNCHWToNHWCPass.cpp]: I've updated the code. Thanks.

File: compiler/luci/pass/src/ConvertNCHWToNHWCPass.cpp [hunk 3]
```diff
@@ -67,6 +67,18 @@ DataFormat get_data_format(loco::Node *node)
...
+bool has_dynamic_shape(const loco::Node *node)
+{
+  const auto circle_node = loco::must_cast<const luci::CircleNode *>(node);
+  const auto shape_signature = circle_node->shape_signature().as_vector();
+  for (auto ss : shape_signature)
+  {
+    if (ss == -1)
+      return true;
+  }
+  return false;
```
Comment (llFreetimell) [inline compiler/luci/pass/src/ConvertNCHWToNHWCPass.cpp:79]: Because of #5501, shape signature will be deprecated.
Please use following codes :)

```suggestion
  for (uint32_t i = 0;i < circle_node->rank(); ++i)
    if (!circle_node->dim(i).known())
      return true;
  return false;
```

Comment (jinevening) [reply compiler/luci/pass/src/ConvertNCHWToNHWCPass.cpp]: Thank you for the instruction :) I've updated the code.

**Review by llFreetimell — APPROVED** [review]
LGTM

**Review by seanshpark — APPROVED** [review]
LGTM

**Review by mhs4670go — APPROVED** [review]
LGTM
